### ISSUE TITLE
[crash] libtorrent::disk_io_thread::call_job_handlers

### ISSUE THREAD

**Issue by pavel-pimenov** [arvidn/libtorrent#1369]
libtorrent:master
https://drdump.com/Problem.aspx?ProblemID=250624&login=guest

![2016-12-01_06-40-07](https://cloud.githubusercontent.com/assets/920705/20781107/5b937ede-b791-11e6-8d6a-303f11793625.png)

![2016-12-01_06-43-26](https://cloud.githubusercontent.com/assets/920705/20781135/84b558e6-b791-11e6-91b6-8ba045bbe99a.png)

**Comment by ssiloti** [comment]
Looking at the code here I think I'm missing something because it looks obviously broken. The disk IO subsystem can not handle reads larger than 16KB but `smart_ban_plugin::on_read_failed_block` tries to read a full piece from the buffer which is not guaranteed to be larger than 16KB.

**Comment by arvidn** [comment]
it's supposed to read all blocks individually, and remember the hash of them and which peer sent them

**Comment by arvidn** [comment]
hm.. yeah, that hash update call seems to be using the piece size, rather than the block size. I bet this is an old bug.

**Comment by pavel-pimenov** [comment]
This is new bug.. 
Crash build: 7.7.503.20146  (src - https://yadi.sk/d/pHQR9_kUzuGJF)
Build 7.7.503.20141 - all ok (src - https://yadi.sk/d/FUw8pRQEzuGZo)

Help?

**Comment by aldenml** [comment]
yes, it's new, right here https://github.com/arvidn/libtorrent/commit/fc9dabe1ab954c80d4c4072063545e3f87308d98#diff-6660250b6a650fff09e1edbe1694e683R189 and I think here too https://github.com/arvidn/libtorrent/commit/fc9dabe1ab954c80d4c4072063545e3f87308d98#diff-6660250b6a650fff09e1edbe1694e683L270

**Comment by aldenml** [comment]
here it is a fix attempt https://github.com/arvidn/libtorrent/pull/1371

**Comment by arvidn** [comment]
merged

**Comment by pavel-pimenov** [comment]
Close issue?
